Change untagged_unions to not allow union fields with drop

bluss opened the pull request:
Change untagged_unions to not allow union fields with drop

Union fields may now never have a type with attached destructor.
This for example allows unions to use arbitrary field types only by wrapping
them in ManuallyDrop (or similar).

The stable rule remains, that union fields must be Copy. We use the new
rule for the `untagged_union` feature.

See RFC 2514, tracking issue #55149

rust-highfive commented:
r? @estebank

(rust_highfive has picked a reviewer for you, use r? to override)

bluss commented:
cc @eddyb @RalfJung

eddyb commented on src/librustc/middle/stability.rs:
```diff
@@ -764,13 +764,15 @@ impl<'a, 'tcx> Visitor<'tcx> for Checker<'a, 'tcx> {
+
+                let param_env = self.tcx.param_env(def_id);
+                let field_needs_drop = adt_def.non_enum_variant().fields.iter().any(|field| {
+                        self.tcx.type_of(field.did).needs_drop(self.tcx, param_env)
```
This is indented too much inwards.

eddyb commented on src/librustc/middle/stability.rs:
```diff
@@ -764,13 +764,15 @@ impl<'a, 'tcx> Visitor<'tcx> for Checker<'a, 'tcx> {
+                }
+
+                let param_env = self.tcx.param_env(def_id);
+                let field_needs_drop = adt_def.non_enum_variant().fields.iter().any(|field| {
```
Name could be `any_fields_need_drop`.

eddyb commented on src/librustc/middle/stability.rs:
```diff
@@ -764,13 +764,15 @@ impl<'a, 'tcx> Visitor<'tcx> for Checker<'a, 'tcx> {
+                });
+                if field_needs_drop {
+                    self.tcx.sess.span_err(item.span, "unions may not contain fields \
+                                                       that need dropping");
```
Note that this is still under `!self.tcx.features().untagged_unions`, so the error won't get triggered with `#![feature(untagged_unions)]`, which is likely not what we want.

Also, it should *probably* not be in this file if it's not tied to stability checks.

cc @nikomatsakis Do you have any suggestions for where to place this code?

RalfJung replied:
This is some form of well-formedness check of a type declaration. Where do we usually do these? Things like "recursive occurrences must be below a pointer indirection", "only the last field of a struct may be unsized", ...

eddyb commented:
r? @nikomatsakis

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/462485192) ([raw log](https://api.travis-ci.org/v3/job/462485192/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:00038e3c:start=1543768431821801159,finish=1543768491192500693,duration=59370699534
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-5.0
---
[00:06:41]    Compiling syntax_ext v0.0.0 (/checkout/src/libsyntax_ext)
[00:07:59] error: unused variable: `ty`
[00:07:59]    --> src/librustc/middle/stability.rs:761:21
[00:07:59]     |
[00:07:59] 761 |                 let ty = self.tcx.type_of(def_id);
[00:07:59]     |                     ^^ help: consider using `_ty` instead
[00:07:59]     = note: `-D unused-variables` implied by `-D warnings`
[00:07:59] 
[00:07:59] error: aborting due to previous error
[00:07:59] 
[00:07:59] 
[00:07:59] error: Could not compile `rustc`.
[00:07:59] 
[00:07:59] To learn more, run the command again with --verbose.
[00:07:59] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json"
[00:07:59] expected success, got: exit code: 101
[00:07:59] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:07:59] Build completed unsuccessfully in 0:04:45
[00:07:59] make: *** [all] Error 1
[00:07:59] Makefile:28: recipe for target 'all' failed
47684 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release
46352 ./src/llvm-emscripten/test/MC
45752 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps
44272 ./src/libcompiler_builtins
---
travis_time:end:1e2aac1c:start=1543768980348625398,finish=1543768980354680610,duration=6055212
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0136f2c9
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:089fb5ca
travis_time:start:089fb5ca
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:027e490b
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented on src/test/run-pass/union/union-manuallydrop.rs:
```diff
@@ -0,0 +1,44 @@
+    // NoDrop should not make needs_drop true
+    assert!(!needs_drop::<UnionOk1<String>>());
+    assert!(!needs_drop::<UnionOk2>());
+    assert!(!needs_drop::<UnionOk3<&Iterator<Item=i32>>>());
```
```suggestion
    assert!(!needs_drop::<UnionOk3<&Iterator<Item = i32>>>());
```

Centril commented on src/test/run-pass/union/union-manuallydrop.rs:
```diff
@@ -0,0 +1,44 @@
+}
+
+fn main() {
+    // NoDrop should not make needs_drop true
```
```suggestion
    // NeedDrop should not make needs_drop true
```

Centril commented on src/test/run-pass/union/union-manuallydrop.rs:
```diff
@@ -0,0 +1,44 @@
+}
+
+#[allow(dead_code)]
+union UnionOk3<T: Copy> {
```
It would also be good to add a test that checks how the compiler handles:
```rust
trait ImpliesCopy: Copy {}

union UnionOk4<T: ImpliesCopy> {
    empty: (),
    value: T,
}
```

A test file checking that `T :/ Copy` will lead to an error would also be good.

bluss replied:
The ui tests touched in the PR cover the non-Copy case, but do we need another test?

bluss replied:
Ok, can't have a failing case in run-pass. It looks like this is normally in ui tests now, so I can add it there (I can't see a single compile-fail test for unions).

Centril replied:
UI tests are afaik compile-fail by default; but yeah, `U5` needs to be added to a different file which has `// compile-fail`.

Centril commented on src/test/run-pass/union/union-manuallydrop.rs:
```diff
@@ -0,0 +1,44 @@
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
```
Copyright is not needed.

RalfJung commented:
> As written, this is insta-stable, but we might want to change that. Nightly with the untagged_unions feature gate doesn't have the requirement we're loosening here, so for users of that feature, nothing changes.

I'd suggest we change the behavior of the feature gate only. So behavior of stable does not change, but we can test on nightly how these unions behave (e.g. the behavior of drop on assignment and the initialization checker might have to be tweaked to match the RFC).

My expectation would be that nightly with the `untagged_union` feature gate has this requirement.

bluss commented:
@RalfJung That makes sense, so we'd make a breaking change for that feature gate. In that case there should definitely be a suggestion to use ManuallyDrop as a diagnostic too.

bluss commented:
@nikomatsakis Any thoughts on what the correct place in the code is for this, a check union fields don't need dropping?

nikomatsakis commented:
@bluss I think this check would make sense to do in this function

https://github.com/rust-lang/rust/blob/8375ab4ff43474c73e3572c2b226560f8cc8e695/src/librustc_typeck/check/mod.rs#L1303-L1312

bluss commented:
Ok, updated with the new rule:

- stable: unchanged
- nightly/untagged_union: don't allow union fields with drop

I've fixed tests quickly but with the best sense I had, so those could require ideas in review.
The first crate I've seen break on the changed rule for `untagged_union` is smallvec, so build will not succeed until crates we depend on are fixed; unsure which crates exactly yet.

bluss commented on src/librustc_typeck/diagnostics.rs:
```diff
@@ -4826,6 +4826,10 @@ type, it's not allowed to override anything in those implementations, as it
+E0720: r##"
+A `union` can not have fields with destructors.
+"##,
+
```
Do we have a rule behind how to assign error numbers? Yes, the text is a bit lacking here.

nikomatsakis replied:
We do not really have a system, no. You just pick one.

bluss commented on src/test/ui/union/union-unsafe.rs:
```diff
@@ -42,14 +46,20 @@ fn main() {
+
+    let mut u3 = U3 { a: ManuallyDrop::new(String::from("old")) }; // OK
+    u3.a = ManuallyDrop::new(String::from("new")); //~ ERROR assignment to non-`Copy` union
+    *u3.a = String::from("new"); //~ ERROR access to union field is unsafe
```
(Above 7 lines).

The different rules between `ManuallyDrop<i32>` vs `ManuallyDrop<String>` are a bit confusing, but due to the Copy/non-Copy distinction.

nikomatsakis replied:
To be clear, this distinction is pre-existing, right? Or are you referring to some behavior that this RFC introduced?

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/470195601) ([raw log](https://api.travis-ci.org/v3/job/470195601/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:09eb2ffc:start=1545250379263888141,finish=1545250478781955081,duration=99518066940
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:19:03]    Compiling ena v0.11.0
[00:19:04]    Compiling smallvec v0.6.7
[00:19:04]    Compiling lock_api v0.1.3
[00:19:04]    Compiling crossbeam-epoch v0.3.1
[00:19:04] error[E0720]: unions may not contain fields that need dropping
[00:19:04]     |
[00:19:04]     |
[00:19:04] 272 | / union SmallVecData<A: Array> {
[00:19:04] 273 | |     inline: A,
[00:19:04] 274 | |     heap: (*mut A::Item, usize),
[00:19:04] 275 | | }
[00:19:04]     | |_^ ManuallyDrop can be used to wrap such fields
[00:19:04] error: aborting due to previous error
[00:19:04] 
[00:19:04] For more information about this error, try `rustc --explain E0720`.
[00:19:04] error: Could not compile `smallvec`.
[00:19:04] error: Could not compile `smallvec`.
[00:19:04] warning: build failed, waiting for other jobs to finish...
[00:19:05] error: build failed
[00:19:05] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json"
[00:19:05] expected success, got: exit code: 101
[00:19:05] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:19:05] Build completed unsuccessfully in 0:16:02
[00:19:05] Makefile:28: recipe for target 'all' failed
[00:19:05] make: *** [all] Error 1
travis_time:end:0079be01:start=1545251632995024654,finish=1545251633007078065,duration=12053411
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:0b37c5c0
travis_time:start:0b37c5c0
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:0b37c5c0:start=1545251633011693449,finish=1545251633016395192,duration=4701743
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0976b9a2
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:1423dc24
travis_time:start:1423dc24
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No s
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/470222106) ([raw log](https://api.travis-ci.org/v3/job/470222106/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:247d7f06:start=1545254645586090975,finish=1545254726658520552,duration=81072429577
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:10:29]    Compiling rustc_typeck v0.0.0 (/checkout/src/librustc_typeck)
[00:10:32] error[E0615]: attempted to take value of method `hir` on type `rustc::ty::TyCtxt<'a, 'tcx, 'tcx>`
[00:10:32]     --> src/librustc_typeck/check/mod.rs:1331:38
[00:10:32]      |
[00:10:32] 1331 |                 let field_span = tcx.hir.span_if_local(field.did).unwrap();
[00:10:32]      |
[00:10:32]      = help: maybe a `()` to call it is missing?
[00:10:32] 
[00:10:35] error: aborting due to previous error
---
travis_time:end:0974e7f2:start=1545255518727292935,finish=1545255518731754954,duration=4462019
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:04ad2652
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:02080aa0
travis_time:start:02080aa0
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bluss commented:
(Oh the latest change with an error about accessing hir shows that I need to rebase to a newer rustc. I'll get back to this.)

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/471184571) ([raw log](https://api.travis-ci.org/v3/job/471184571/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0b8bf000:start=1545431488154224782,finish=1545431489878503383,duration=1724278601
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
travis_time:end:0760e260:start=1545432986152532478,finish=1545432986160581650,duration=8049172
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:07377628
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:00fd52a0
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/471284942) ([raw log](https://api.travis-ci.org/v3/job/471284942/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0c5a3ef7:start=1545467341482658400,finish=1545467398630855457,duration=57148197057
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
187084 ./obj/build/cache/2018-12-09
160388 ./obj/build/bootstrap/debug/incremental
153276 ./src/tools/clang
144288 ./obj/build/bootstrap/debug/incremental/bootstrap-2x7szixskz2uj
144284 ./obj/build/bootstrap/debug/incremental/bootstrap-2x7szixskz2uj/s-f7tlz0n17y-126lqir-3g9jcdhvsk79m
113760 ./obj/build/x86_64-unknown-linux-gnu/stage1
113740 ./obj/build/x86_64-unknown-linux-gnu/stage1/lib
111164 ./obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends
107420 ./src/tools/lldb
---
travis_time:end:140f6e15:start=1545468873316874322,finish=1545468873321514721,duration=4640399
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:27e96dd8
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:00870e64
travis_time:start:00870e64
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/471296262) ([raw log](https://api.travis-ci.org/v3/job/471296262/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:00bfe1ef:start=1545472031155931615,finish=1545472088481215912,duration=57325284297
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:01:13] 
######################################################################## 100.0%
[00:01:13] extracting /checkout/obj/build/cache/2018-12-09/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:01:13]     Updating crates.io index
[00:01:21]     Updating git repository `https://github.com/servo/rust-smallvec`
[00:01:21] error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
[00:01:21] Build completed unsuccessfully in 0:00:19
[00:01:21] Makefile:81: recipe for target 'prepare' failed
[00:01:21] make: *** [prepare] Error 1
[00:01:22] Command failed. Attempt 2/5:
[00:01:22] Command failed. Attempt 2/5:
[00:01:23] error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
[00:01:23] Build completed unsuccessfully in 0:00:00
[00:01:23] Makefile:81: recipe for target 'prepare' failed
[00:01:23] make: *** [prepare] Error 1
[00:01:25] Command failed. Attempt 3/5:
[00:01:25] Command failed. Attempt 3/5:
[00:01:25] error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
[00:01:25] Build completed unsuccessfully in 0:00:00
[00:01:25] Makefile:81: recipe for target 'prepare' failed
[00:01:25] make: *** [prepare] Error 1
[00:01:28] Command failed. Attempt 4/5:
[00:01:28] Command failed. Attempt 4/5:
[00:01:28] error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
[00:01:28] Build completed unsuccessfully in 0:00:00
[00:01:28] Makefile:81: recipe for target 'prepare' failed
[00:01:28] make: *** [prepare] Error 1
[00:01:32] Command failed. Attempt 5/5:
[00:01:32] Command failed. Attempt 5/5:
[00:01:33] error: the lock file /checkout/Cargo.lock needs to be updated but --locked was passed to prevent this
[00:01:33] Build completed unsuccessfully in 0:00:00
[00:01:33] Makefile:81: recipe for target 'prepare' failed
[00:01:33] make: *** [prepare] Error 1
[00:01:33] The command has failed after 5 attempts.
---
travis_time:end:1358601c:start=1545472190508279787,finish=1545472190514602852,duration=6323065
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0a4bf021
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:025c5ae8
travis_time:start:025c5ae8
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:181c644c
$ dmesg | grep -i kill
$ dmesg | grep -i kill
[   10.762600] init: failsafe main process (1094) killed by TERM signal
[   42.170072] init: plymouth-upstart-bridge main process (510) killed by TERM signal
travis_fold:end:after_failure.6

Done. Your build exited with 1.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bluss commented:
Hrm, the smallvec from git experiment (just to have travis run all tests) failed, so I'll just wait for a release of that crate.

bluss commented:
@RalfJung If I understand you correctly, uninitialized data inside a ManuallyDrop is a no-no while inside a union in might be ok.

Therefore the following transition (forced by this change):

```
union Foo<T> {
   value:T
}
```
becomes
```
union Foo<T> {
   value: ManuallyDrop<T>
}
```

is a hazard, and we may be introducing new issues with soundness?

RalfJung commented:
@bluss Your second union is equivalent to `MaybeUninit<T>`. `MaybeUninit` is defined using `ManuallyDrop`.

The inside of a union does not have to be initialized, it is just ignored.  Hence a `ManuallyDrop` inside a union does not have to be initialized either.  Only when you actually access that field as a `ManuallyDrop` are you making any promise. 

Basically, whenever you access/copy data at some type, it must be valid (initialized) for that type. This validity check proceeds recursively through structs (including tuples, arrays, whatever) and enums (taking the current discriminant into account). When, during the validity check, you encounter a union, this branch of the recursive search just stops because there is nothing to check. Does that make sense?

bluss commented:
@RalfJung Thanks, ok that relieves some of the worry.

Access sounds nebulous when it comes to partial initialization:

```rust
union Foo<T> {
    empty: (),
    value: ManuallyDrop<[T; 2]>
}
```

I propose sound code would handle this by:

```rust
let mut data = Foo { empty: () };

// initialize first element
let first_elem_1 = &mut data.value as *mut _ as *mut T;
ptr::write(first_elem_1, T::default());

// read first element
let first_elem_2 = &data.value as *const _ as *const T;
println!("{}", *first_elem_2);
```

Is this treatment with raw pointers enough so that the validity check only concerns the value of `T` and not the `[T; 2]`? The array only being partially initialized, so the `[T; 2]` value is not valid and the `ManuallyDrop<[T; 2]>` value is not valid.

RalfJung commented:
@bluss Now you are running into https://github.com/rust-lang/rfcs/pull/2582. ;)

`&mut data.value` creates a `&mut ManuallyDrop<[T; 2]>` which must be valid. That may or may not require the data behind the pointer to be valid -- I think it should not make such requirements, but the discussion still has to happen in the UCG group (and then an RFC). But either way, if we adapt the aforementioned RFC you'd not even create a reference and hence you'd be good.

Still I'd prefer if you didn't have to cast around raw pointers like that, that's a very dangerous thing to do.

bluss commented:
Great, maybe I should talk in that rfc so that it is explicitly mentioned.

AFAIK, there is no guarantee we can cast the raw-ptr-to-union to a raw ptr of one of its fields, but maybe that's implicit somewhere. There's no repr(transparent) for unions.

bors commented:
:umbrella: The latest upstream changes (presumably #57108) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@bluss @RalfJung is this code ready for review then? what is current status?

RalfJung commented:
I haven't been following very closely as I don't understand much about this part of the compiler.

nikomatsakis commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -1311,9 +1311,40 @@ fn check_union<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                // We are currently checking the type this field came from, so it must be local
+                let field_span = tcx.hir().span_if_local(field.did).unwrap();
+                let param_env = tcx.param_env(field.did);
+                if field_ty.needs_drop(tcx, param_env) {
```
Hmm. In general, the `needs_drop` accessor is a "heuristic", basically used for generating more efficient MIR. I guess though that it is relevant to our semantics in terms of the NLL borrow check as well. I think I might prefer if we made a "language definition" version of this function at *some point* -- or maybe for now it suffices to amend the [rustc documentation for this function](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.needs_drop) to note that it is used for checking union fields and the like.

bluss replied:
Great, updated docs. Yes, it seems like this method is sliding towards having to be more exact.

nikomatsakis commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -1311,9 +1311,40 @@ fn check_union<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                if field_ty.needs_drop(tcx, param_env) {
+                    struct_span_err!(tcx.sess, field_span, E0720,
+                                     "unions may not contain fields that need dropping")
+                                .span_note(field_span, "ManuallyDrop can be used to wrap the type")
```
style note: we should mark this with backticks like so

```
`ManuallyDrop`
```

Perhaps better to write `std::mem::ManuallyDrop`, as well.

bluss replied:
Thanks, updated

nikomatsakis commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -1311,9 +1311,40 @@ fn check_union<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                let param_env = tcx.param_env(field.did);
+                if field_ty.needs_drop(tcx, param_env) {
+                    struct_span_err!(tcx.sess, field_span, E0720,
+                                     "unions may not contain fields that need dropping")
```
Eventually, it would be nice if we could say with more detail *which type* needs to be dropped

nikomatsakis replied:
but I guess it doesn't have to be part of this PR

bluss commented:
> To be clear, this distinction is pre-existing, right? Or are you referring to some behavior that this RFC introduced?

(For some reason github didn'tt let me reply to this comment inline.) Yes, it's preexisting, it just gets a bit interesting with ManuallyDrop. The old rules for Copy to assign don't seem to make as much sense.

Thanks for the comments, will get going with fixing the test

bluss commented:
Rebased & update for review comments (documentation and diagnostic note formatting about ManuallyDrop)

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/170050534) ([raw log](https://api.travis-ci.com/v3/job/170050534/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:2232377a:start=1547394252117735667,finish=1547394321336588275,duration=69218852608
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
tidy check
[00:03:15] * 570 error codes
[00:03:15] * highest error code: E0730
[00:03:15] * 246 features
[00:03:16] Stray file with UI testing output: "/checkout/src/test/ui/union/union-with-drop-fields-lint.stderr"
[00:03:16] some tidy checks failed
[00:03:16] 
[00:03:16] 
[00:03:16] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:03:16] 
[00:03:16] 
[00:03:16] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:03:16] Build completed unsuccessfully in 0:00:46
[00:03:16] Build completed unsuccessfully in 0:00:46
[00:03:16] Makefile:69: recipe for target 'tidy' failed
[00:03:16] make: *** [tidy] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:07f28b13
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Sun Jan 13 15:48:45 UTC 2019
---
travis_time:end:11f1ad85:start=1547394526559604893,finish=1547394526563887815,duration=4282922
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0001974c
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:12cdc67a
travis_time:start:12cdc67a
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:043c7e24
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/170261598) ([raw log](https://api.travis-ci.com/v3/job/170261598/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:16660f88:start=1547485910734037882,finish=1547486025282440459,duration=114548402577
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:30:55] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json"
[00:30:55] expected success, got: exit code: 101
[00:30:55] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:30:55] Build completed unsuccessfully in 0:27:07
[00:30:55] Makefile:18: recipe for target 'all' failed
[00:30:55] make: *** [all] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:08ba204a
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon Jan 14 17:44:52 UTC 2019
---
travis_time:end:0c11cea9:start=1547487893216508007,finish=1547487893221420257,duration=4912250
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0ad077e0
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:11526274
travis_time:start:11526274
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:2219d9ff
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bluss commented:
Sure, I wasn't sure about when to request a new smallvec release, but I've done so now.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/171645354) ([raw log](https://api.travis-ci.com/v3/job/171645354/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:054b782a:start=1547942650229009889,finish=1547942721247305906,duration=71018296017
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:28:55]    Compiling arena v0.0.0 (/checkout/src/libarena)
[00:28:55]    Compiling syntax_pos v0.0.0 (/checkout/src/libsyntax_pos)
[00:29:01]    Compiling rustc_errors v0.0.0 (/checkout/src/librustc_errors)
[00:30:35]    Compiling syntax_ext v0.0.0 (/checkout/src/libsyntax_ext)
[00:30:45] error: internal compiler error: src/librustc/ty/relate.rs:703: impossible case reached: can't relate: '_#0r with Type(RustaceansAreAwesome)
[00:30:45] thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:589:9
[00:30:45] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[00:30:46] error: aborting due to previous error
[00:30:46] 
[00:30:46] 
[00:30:46] 
[00:30:46] note: the compiler unexpectedly panicked. this is a bug.
[00:30:46] 
[00:30:46] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[00:30:46] 
[00:30:46] note: rustc 1.33.0-dev running on x86_64-unknown-linux-gnu
[00:30:46] 
[00:30:46] note: compiler flags: -Z force-unstable-if-unmarked -C prefer-dynamic -C opt-level=2 -C prefer-dynamic -C debug-assertions=y -C link-args=-Wl,-rpath,$ORIGIN/../lib --crate-type dylib
[00:30:46] note: some of the compiler flags provided by cargo are hidden
[00:30:46] 
[00:30:46] error: Could not compile `rustc`.
[00:30:46] warning: build failed, waiting for other jobs to finish...
[00:30:46] warning: build failed, waiting for other jobs to finish...
[00:31:05] error: build failed
[00:31:05] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json"
[00:31:05] expected success, got: exit code: 101
[00:31:05] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:31:05] Build completed unsuccessfully in 0:27:27
[00:31:05] make: *** [all] Error 1
[00:31:05] Makefile:18: recipe for target 'all' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0eac3d5b
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Sun Jan 20 00:36:35 UTC 2019
---
travis_time:end:13a8477a:start=1547944595839519385,finish=1547944595844280318,duration=4760933
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0001ba58
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:2f571240
travis_time:start:2f571240
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:08a87700
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bluss commented:
The ICE reproduces locally, same message as #57156 and a smallvec in the backtrace, but will have to look closer at it.

bors commented:
:umbrella: The latest upstream changes (presumably #57805) made this pull request unmergeable. Please resolve the merge conflicts.

TimNN commented:
Ping from triage @bluss: What is the status of this PR?

Dylan-DPC commented:
ping from triage @bluss any updates on this?

Dylan-DPC commented:
ping from triage @bluss 
Unfortunately we haven't heard from you on this in a while, so I'm closing the PR to keep things tidy. Don't worry though, if you'll have time again in the future please reopen this PR, we'll be happy to review it again!
